google-cloud-core/src/main/java/com/google/cloud/ServiceOptions.java

```diff
@@ -394,8 +394,11 @@ protected static String getAppEngineProjectId() {
       String serviceAccountName = (String) method.invoke(appIdentityService);
       int indexOfAtSign = serviceAccountName.indexOf('@');
       return serviceAccountName.substring(0, indexOfAtSign);
-    } catch (Exception ignore) {
-      // return null if can't determine
+    } catch (Exception exception) {
+      if (System.getProperty("com.google.appengine.runtime.version") != null) {
+        // Could not resolve appengine classes under GAE environment.
+        throw new RuntimeException("Unable to resolve appengine classes. Missing dependency.");
```

Comment:
I was thinking more of a getting-started section for App Engine somewhere (maybe even in the main README.md file, or possibly a new APPENGINE.md file), not a troubleshooting section. The message should be something like "Unable to resolve App Engine classes. See http://.../...md for details on using google-cloud-java on App Engine."